test/assignment.coffee

```diff
@@ -947,64 +947,41 @@ test "#4673: complex destructured object spread variables", ->
   eq g, 1
 
 test "#4878: Compile error when using destructuring with a splat or expansion in an array", ->
-  eqJS '''
-    foo = (list) ->
-      [first, ..., last] = list
-    ''', '''
-    var foo,
-      slice = [].slice;
-
-    foo = function(list) {
-      var first, last;
-      return [first] = list, [last] = slice.call(list, -1);
-    };
-    '''
-
-  eqJS '''
-    foo = (list) ->
-      [first..., last] = list
-    ''', '''
-    var foo,
-      splice = [].splice;
-
-    foo = function(list) {
-      var first, last;
-      return [...first] = list, [last] = splice.call(first, -1);
-    };
-    '''
-
-  eqJS '''
-    foo = (list) ->
-      ret =
-        if value.length
-          [ first, ..., last ] = value
-          [ first, last ]
-        else
-          []
-    ''', '''
-    var foo,
-      slice = [].slice;
-
-    foo = function(list) {
-      var first, last, ret;
-      return ret = value.length ? (([first] = value, [last] = slice.call(value, -1)), [first, last]) : [];
-    };
-    '''
-
-  eqJS '''
-    foo = (list) ->
-      ret =
-        if value.length
-          [ first..., last ] = value
-          [ first, last ]
-        else
-          []
-    ''', '''
-    var foo,
-      splice = [].splice;
-
-    foo = function(list) {
-      var first, last, ret;
-      return ret = value.length ? (([...first] = value, [last] = splice.call(first, -1)), [first, last]) : [];
-    };
-    '''
+  arr = ['a', 'b', 'c', 'd']
+
+  f1 = (list) ->
+    [first, ..., last] = list
+
+  f2 = (list) ->
+    [first..., last] = list
+
+  f3 = (list) ->
+    ([first, ...] = list); first
+
+  f4 = (list) ->
+    ([first, ...rest] = list); rest
+
+  arrayEq f1(arr), ['d']
+  arrayEq f2(arr), ['d']
+  arrayEq f3(arr), 'a'
+  arrayEq f4(arr), ['b', 'c', 'd']
+
+  foo = (list) ->
+    ret =
+      if list.length
```

Comment:
I don’t like using `list.length` being nonzero as a proxy for truthiness. If you’re evaluating that `list.length isnt 0`, you should be explicit about it. I think it’s a lot more readable than just `list.length` or other hacks like `!!`.

Reply:
Right. I copied the example from #4878.